libcaf_core/caf/thread_hook.hpp

```diff
@@ -0,0 +1,48 @@
+/******************************************************************************
+ *                       ____    _    _____                                   *
+ *                      / ___|  / \  |  ___|    C++                           *
+ *                     | |     / _ \ | |_       Actor                         *
+ *                     | |___ / ___ \|  _|      Framework                     *
+ *                      \____/_/   \_|_|                                      *
+ *                                                                            *
+ * Copyright (C) 2011 - 2017                                                  *
+ * Dominik Charousset <dominik.charousset (at) haw-hamburg.de>                *
+ *                                                                            *
+ * Distributed under the terms and conditions of the BSD 3-Clause License or  *
+ * (at your option) under the terms and conditions of the Boost Software      *
+ * License 1.0. See accompanying files LICENSE and LICENSE_ALTERNATIVE.       *
+ *                                                                            *
+ * If you did not receive a copy of the license files, see                    *
+ * http://opensource.org/licenses/BSD-3-Clause and                            *
+ * http://www.boost.org/LICENSE_1_0.txt.                                      *
+ ******************************************************************************/
+
+#ifndef CAF_THREAD_HOOK_HPP
+#define CAF_THREAD_HOOK_HPP
+
+namespace caf {
+
+/// Interface to define thread hooks.
+class thread_hook {
+public:
+  virtual ~thread_hook() {
+    // nop
+  }
+
+  /// Called by the actor system once before starting any threads.
+  virtual void init(actor_system&) = 0;
+
+  /// Called whenever the actor system has started a new thread.
+  /// To access a reference to the started thread use `std::this_thread`.
+  /// @warning must the thread-safe
+  virtual void thread_started_cb(actor_system&) = 0;
```

Comment:
The `actor_system&` argument is redundant, because of the `init` member function. Any implementation can simply store a pointer to the actor system and than have all the context available at any time.

Additionally, I would strip the `_cb` suffix. You've probably adapted this from the `io::hook` class, but the only reason the suffix is used in that class is to avoid name clashes with the enum values of `io::hook::event_type`.